### PR TITLE
Cdebug

### PR DESCRIPTION
**Pull request by CaptBoscho**
rolling a 7 better code
wait for Derek's merge. (monday 3:50p.m.)

### CODE REVIEW

File: java/src/server/facade/MockFacade.java [hunk 1]
```diff
@@ -298,6 +298,7 @@ public CommandExecutionResult buildCity(int gameID, int player, VertexLocation l
      */
     @Override
     public CommandExecutionResult offerTrade(int gameID, OfferTradeDTO dto) throws OfferTradeException {
+
```
Comment (dannyharding10) [inline java/src/server/facade/MockFacade.java]: This is unnecessary

File: java/src/shared/model/game/Game.java [hunk 2]
```diff
@@ -813,6 +813,15 @@ public void discardCards(int playerIndex, List<ResourceType> cards) throws Playe
         if (canDiscardCards(playerIndex) && this.turnTracker.canDiscard()) {
             playerManager.discardResourceType(playerIndex, cards);
         }
+        boolean change = true;
+        for(Player p : getPlayers()){
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: can you call this player?

Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: I'm awesome and i don't care what anyone says I'm too cool. hollaaaaaaa

File: java/src/shared/model/game/Game.java [hunk 3]
```diff
@@ -813,6 +813,15 @@ public void discardCards(int playerIndex, List<ResourceType> cards) throws Playe
         if (canDiscardCards(playerIndex) && this.turnTracker.canDiscard()) {
             playerManager.discardResourceType(playerIndex, cards);
         }
+        boolean change = true;
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: This will be more readable if you call it `allPlayersDiscarded` or something like that

File: java/src/shared/model/game/Game.java [hunk 4]
```diff
@@ -826,13 +835,22 @@ public void rollNumber(final int value) throws Exception {
         //Is value a 7 - robber
         if (value == 7) {
             //Go to discarding phase before robbing if any player has to discard
-            getPlayers().forEach(player -> {
+            boolean discard = false;
+            for(Player player: getPlayers()){
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: thanks for making this a normal for loop!  Much more readable

File: java/src/shared/model/game/Game.java [hunk 5]
```diff
@@ -826,13 +835,22 @@ public void rollNumber(final int value) throws Exception {
...
                 if (player.canDiscardCards()) {
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: This needs to have another condition, `!player.hasDiscarded()`.  otherwise a player with 20 cards will have to discard twice.   `setDiscarded(false)` will have to be called for each player right when a 7 is rolled.

Comment (CaptBoscho) [inline java/src/shared/model/game/Game.java]: Danny, This function is only called when the 7 is rolled, which means we went encounter that error.

File: java/src/shared/model/game/Game.java [hunk 6]
```diff
@@ -826,13 +834,22 @@ public void rollNumber(final int value) throws Exception {
...
-            });
+            }
+            if(discard){
+               turnTracker.setPhase(TurnTracker.Phase.DISCARDING);
+            }else{
+                turnTracker.setPhase(TurnTracker.Phase.ROBBING);
+            }
             //Otherwise just move to the robbing phase
-            turnTracker.setPhase(TurnTracker.Phase.ROBBING);
+            //Can't move to robbing phase here! need to wait for everyone to discard
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: both comments here can be taken out

File: java/src/shared/model/game/Game.java [hunk 7]
```diff
@@ -845,7 +862,9 @@ public void rollNumber(final int value) throws Exception {
             });
 
             //Move to next phase - Playing
-            turnTracker.nextPhase();
+            //turnTracker.nextPhase();
```
Comment (dannyharding10) [inline java/src/shared/model/game/Game.java]: if we're not using this code you can just delete the line

File: java/test/server/commands/moves/OfferTradeCommandTest.java [hunk 8]
```diff
@@ -23,8 +32,21 @@ public void tearDown() {
...
+        TradePackage package1 = new TradePackage(0,send);
+        TradePackage package2 = new TradePackage(1,receive);
+        Trade trade = new Trade(package1,package2);
+        OfferTradeDTO dto = new OfferTradeDTO(0,trade,1);
+        try {
+            CommandExecutionResult answer = Config.facade.offerTrade(1,dto);
+        } catch (OfferTradeException e) {
+            e.printStackTrace();
+        }
     }
```
Comment (dannyharding10) [inline java/test/server/commands/moves/OfferTradeCommandTest.java]: Awesome! looks like a good test.  This could potentially be in another pull request, so they're split up and smaller.

File: java/src/shared/model/game/Game.java (hunk 4 above)
Comment (derekargueta) [inline java/src/shared/model/game/Game.java]: :(

File: java/test/server/commands/moves/OfferTradeCommandTest.java (hunk 8 above)
Comment (derekargueta) [inline java/test/server/commands/moves/OfferTradeCommandTest.java]: Can you add a few more test cases? This only tests the straight-forwarded expected norm case. Please throw in some `null`s, negative numbers, big numbers, etc. to thoroughly test the logic.

Comment (derekargueta) [inline java/test/server/commands/moves/OfferTradeCommandTest.java]: Also this isn't actually a test..... there's no JUnit assertions.....

File: java/src/shared/model/game/Game.java (hunk 5 above)
Comment (yeyyosing) [inline java/src/shared/model/game/Game.java]: besides, the client makes it impossible to discard twice anyway.  so you are good corbin
